test/exe/main_common_test.cc

```diff
@@ -104,6 +104,31 @@ TEST_P(MainCommonTest, ConstructDestructHotRestartDisabledNoInit) {
   EXPECT_TRUE(main_common.run());
 }
 
+// This test verifies that validation can run from MainCommonBase, from a thread
+// other than the test thread. This is a desirable calling sequence in some
+// contexts. To make this work we must declare MainThread in MainCommonBase, in
+// addition to declaring it in MainCommon. There is no harm in double-declaring.
+TEST_P(MainCommonTest, ValidateUsingMainCommonBaseOutsideTestThread) {
+  EXPECT_FALSE(Thread::MainThread::isMainThreadActive());
+  {
+    Event::TestRealTimeSystem real_time_system;
+    DefaultListenerHooks default_listener_hooks;
+    ProdComponentFactory prod_component_factory;
```

Comment:
can we push these down to where they're used e.g. just above the `MainCommonBase` ctor?

Reply:
done